drivers/counter/counter_ll_stm32_timer.c

```diff
@@ -413,17 +412,28 @@ static int counter_stm32_init_timer(const struct device *dev)
 	cfg->irq_config_func(dev);
 
 	/* initialize timer */
-	LL_TIM_StructInit(&init);
+	LL_TIM_SetPrescaler(timer, cfg->prescaler);
 
-	init.Prescaler = cfg->prescaler;
-	init.CounterMode = LL_TIM_COUNTERMODE_UP;
-	init.Autoreload = counter_get_max_top_value(dev);
-	init.ClockDivision = LL_TIM_CLOCKDIVISION_DIV1;
+	if (IS_TIM_COUNTER_MODE_SELECT_INSTANCE(timer)) {
+		LL_TIM_SetCounterMode(timer, LL_TIM_COUNTERMODE_UP);
+	}
+
+	LL_TIM_SetAutoReload(timer, counter_get_max_top_value(dev));
+
+	if (IS_TIM_CLOCK_DIVISION_INSTANCE(timer)) {
+		LL_TIM_SetClockDivision(timer, LL_TIM_CLOCKDIVISION_DIV1);
+	}
```

Comment:
Nit: I'd group all conditional blocks together
```suggestion
	LL_TIM_SetAutoReload(timer, counter_get_max_top_value(dev));
	
	if (IS_TIM_COUNTER_MODE_SELECT_INSTANCE(timer)) {
		LL_TIM_SetCounterMode(timer, LL_TIM_COUNTERMODE_UP);
	}

	if (IS_TIM_CLOCK_DIVISION_INSTANCE(timer)) {
		LL_TIM_SetClockDivision(timer, LL_TIM_CLOCKDIVISION_DIV1);
	}
```